lib/broadway/producer.ex

```diff
@@ -171,14 +187,63 @@ defmodule Broadway.Producer do
   end
 
   @impl true
+  def handle_info({__MODULE__, :cancel_consumers}, %{rate_limiting: %{} = rate_limiting} = state) do
+    rate_limiting =
+      if :queue.is_empty(rate_limiting.message_buffer) do
+        cancel_consumers(state)
+        rate_limiting
+      else
+        %{rate_limiting | draining?: true}
+      end
+
+    {:noreply, [], %{state | rate_limiting: rate_limiting}}
+  end
+
   def handle_info({__MODULE__, :cancel_consumers}, state) do
-    for from <- state.consumers do
-      send(self(), {:"$gen_producer", from, {:cancel, :shutdown}})
+    cancel_consumers(state)
+    {:noreply, [], state}
+  end
+
+  def handle_info({__MODULE__, :handle_next_demand}, state) do
```

Comment:
So this clause (and therefore handle_demand) can be called when draining, which is not supposed to happen, since demand was set to accumulate. Maybe the best idea is actually to move the check if it the rate_limiter is open and not draining to the function clause and we simply always schedule it whenever the demand buffer is not empty?

In other words, we need to check the state when we actually execute things, not when we schedule them, because it may change since then.